### PR TITLE
fix(docker): use a HEAD request to the real resource auth.

### LINKED ISSUE #14708: ACR Datasource unauthorized while fetching tags list.
### How are you running Renovate?

Self-hosted

### If you're self-hosting Renovate, tell us what version of Renovate you run.

32.6.0 

### Please select which platform you are using if self-hosting.

GitLab self-hosted

### If you're self-hosting Renovate, tell us what version of the platform you run.

14.8.2

### Was this something which used to work for you, and then stopped?

I never saw this working

### Describe the bug

Using ACR. Fetching new docker image tags is failing.
```
DEBUG: Datasource unauthorized (repository=MY/repository)
       "datasource": "docker",
       "packageName": "redacted.azurecr.io/upstream/rust",
       "url": "https://redacted.azurecr.io/v2/upstream/rust/tags/list?n=10000"
       
```
I tried to reproduced the call and authentication performed by renovate.
I'm facing the same authentication issue.

```
{
    "errors": [
        {
            "code": "UNAUTHORIZED",
            "message": "authentication required, visit https://aka.ms/acr/authorization for more information.",
            "detail": [
                {
                    "Type": "repository",
                    "Name": "upstream/rust",
                    "Action": "metadata_read"
                }
            ]
        }
    ]
}
```
Renovate is requesting a token with the following scope. `repository:upstream/rust:pull`.

Scope should be either `repository:upstream/rust:pull,metadata_read` or `repository:upstream/rust:*`

### Relevant debug logs

<details><summary>Logs</summary>

```
Copy/paste any log here, between the starting and ending backticks
```

</details>

### Have you created a minimal reproduction repository?

No reproduction repository

### PR DESCRIPTION
**Pull request by vlagorce**
## Changes

 When listing tags we should make a HEAD request to the real resource and use the auth header values to get a token

## Context

Closes #14708

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [ ] Code inspection only, or
- [x] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

**Comment by rarkins** [comment]
Mistaken package.json change?

**Comment by vlagorce** [comment]
> Mistaken package.json change?

removed.

File: lib/modules/datasource/docker/index.spec.ts [hunk 1]
```diff
@@ -219,7 +255,7 @@ describe('modules/datasource/docker/index', () => {
         .get('/')
         .reply(401, '', {
           'www-authenticate':
-            'Bearer realm="https://auth.docker.io/token",service="registry.docker.io",scope="repository:samalba/my-app:pull  "',
+            'Bearer realm="https://auth.docker.io/token",service="registry.docker.io",scope="repository:library/some-dep:pull  "',
```
Comment (viceice) [inline lib/modules/datasource/docker/index.spec.ts:258]: Can you please revert those unrelated naming changes? Or are the tests wrong?

**Review by viceice — CHANGES_REQUESTED** [review]

File: lib/modules/datasource/docker/index.spec.ts (hunk 1 above)
Comment (vlagorce) [reply lib/modules/datasource/docker/index.spec.ts:258]: The code I have added use the scope provided in 401 response.
I had to either change this line or  line 265 `&scope=repository:library/some-dep:pull'` or this one.
they must be aligned.

File: lib/modules/datasource/docker/index.spec.ts [hunk 2]
```diff
@@ -723,7 +759,7 @@ describe('modules/datasource/docker/index', () => {
 
         httpMock
           .scope('https://ecr-proxy.company.com/v2')
-          .get('/')
+          .head('/tags/list')
```
[the remark this answers is not in the document]
Comment (vlagorce) [reply lib/modules/datasource/docker/index.spec.ts:762]: Indeed.

File: lib/modules/datasource/docker/index.spec.ts [hunk 3]
```diff
@@ -14,7 +14,7 @@ const http = new Http(DockerDatasource.id);
 
 jest.mock('@aws-sdk/client-ecr');
 jest.mock('../../../util/host-rules');
-
+jest.setTimeout(500000);
```
[the remark this answers is not in the document]
Comment (vlagorce) [reply lib/modules/datasource/docker/index.spec.ts:17]: I had issued to debug. I forgot this one.
I'm away from computer. Feel free to suggest/push a remove commit.

Comment (viceice) [inline lib/modules/datasource/docker/index.spec.ts:17]:
```suggestion
```

[the remark this answers is not in the document]
Comment (vlagorce) [reply lib/modules/datasource/docker/index.spec.ts]: Thank you

File: lib/modules/datasource/docker/index.spec.ts [hunk 4]
```diff
@@ -14,7 +14,6 @@ const http = new Http(DockerDatasource.id);
 
 jest.mock('@aws-sdk/client-ecr');
 jest.mock('../../../util/host-rules');
-
 type ECR = _AWS.ECR;
```
Comment (viceice) [inline lib/modules/datasource/docker/index.spec.ts:17]:
```suggestion

type ECR = _AWS.ECR;
```

**Review by viceice — APPROVED** [review]

**Comment by viceice** [comment]
I think this needs some real runs against different docker registries 🤔

**Comment by vlagorce** [comment]
Folowing your mesage I ran this branch against a repo having images hosted in ACR. 
Updates are found without any lookup error. :)

**Comment by rslinckx** [comment]
Might this have caused the following errors on gcr.io (and also on eu.gcr.io) since 2 days?:

```
WARN: Invalid registry response
{
  "apiCheckUrl": "https://gcr.io/v2/datadoghq/cluster-agent/tags/list?n=10000",
  "res": {
    "statusCode": 405,
    "body": "",
    "headers": {
      "docker-distribution-api-version": "registry/2.0",
      "content-type": "application/json",
      "content-encoding": "gzip",
      "date": "Fri, 08 Apr 2022 07:47:28 GMT",
      "server": "Docker Registry",
      "cache-control": "private",
      "x-xss-protection": "0",
      "x-frame-options": "SAMEORIGIN",
      "transfer-encoding": "chunked",
      "alt-svc": "h3=\":443\"; ma=2592000,h3-29=\":443\"; ma=2592000,h3-Q050=\":443\"; ma=2592000,h3-Q046=\":443\"; ma=2592000,h3-Q043=\":443\"; ma=2592000,quic=\":443\"; ma=2592000; v=\"46,43\"",
      "connection": "close"
    },
    "authorization": false
  }
}
DEBUG: Failed to get authHeaders for getTags lookup
```
